platform/default/src/mbgl/storage/pmtiles_file_source.cpp

```diff
@@ -0,0 +1,586 @@
+                }
+
+                auto& minzoom = doc["minzoom"];
+
+                if (minzoom.IsString()) {
+                    minzoom.SetInt(std::atoi(minzoom.GetString()));
+                }
+
+                if (!minzoom.IsNumber()) {
+                    minzoom = rapidjson::Value().SetUint(header.min_zoom);
+                }
+
+                doc["minzoom"] = minzoom;
+
+                if (!doc.HasMember("maxzoom")) {
+                    doc.AddMember("maxzoom", rapidjson::Value(), allocator);
+                }
+
+                auto& maxzoom = doc["maxzoom"];
+
+                if (maxzoom.IsString()) {
+                    maxzoom.SetInt(std::atoi(maxzoom.GetString()));
+                }
+
+                if (!maxzoom.IsNumber()) {
+                    maxzoom = rapidjson::Value().SetUint(header.max_zoom);
+                }
+
+                doc["maxzoom"] = maxzoom;
+
+                std::string metadata = serialize(doc);
+                metadata_cache.emplace(url, metadata);
+
+                callback(std::unique_ptr<Response::Error>());
+            };
+
+            if (header.json_metadata_bytes > 0) {
+                Resource resource(Resource::Kind::Source, url);
+                resource.loadingMethod = Resource::LoadingMethod::Network;
+                resource.dataRange = std::make_pair(header.json_metadata_offset,
+                                                    header.json_metadata_offset + header.json_metadata_bytes - 1);
+
+                tasks[req] = getFileSource()->request(resource, [=](const Response& responseMetadata) {
+                    if (responseMetadata.error) {
+                        callback(std::make_unique<Response::Error>(
+                            responseMetadata.error->reason,
+                            std::string("Error fetching PMTiles metadata: ") + responseMetadata.error->message));
+
+                        return;
+                    }
+
+                    std::string data = *responseMetadata.data;
+
+                    if (header.internal_compression == pmtiles::COMPRESSION_GZIP) {
+                        data = util::decompress(data);
+                    }
+
+                    parse_callback(data);
+                });
+
+                return;
+            }
+
+            parse_callback(std::string());
+        });
+    }
+
+    void storeDirectory(const std::string& url,
+                        uint64_t directoryOffset,
+                        uint64_t directoryLength,
+                        const std::string& directoryData) {
+        if (directory_cache.find(url) == directory_cache.end()) {
+            directory_cache.emplace(url, std::map<std::string, std::vector<pmtiles::entryv3>>());
+            directory_cache_control.emplace(url, std::vector<std::string>());
+        }
+
+        std::string directory_cache_key = url + "|" + std::to_string(directoryOffset) + "|" +
```

Comment:
In the JS implementation we include the returned ETag as part of the cache key: https://github.com/protomaps/PMTiles/blob/main/js/index.ts#L635

Looking at this implementation right now it seems like it does not yet support ETags? That may be fine for the initial part, but ideally the reader should work correctly if the PMTiles archive changes on storage without having to restart the app and clear the in-memory directory cache. 

the AWS implementation here: https://github.com/protomaps/PMTiles/blob/main/serverless/aws/src/index.ts#L79 has examples of how every non-root request sends an `If-Match` header; the JS implementation also permits one and only one retry of any request if the ETag does not match (`412 Precondition Failed`). 

Maybe this feature is best suited for a follow-up PR, but it may be useful to prepare any HTTP request interfaces with ETag in the request/response function signatures. 

Thanks for taking on this feature 👍

Reply:
Nice! I was a little confused about it, but you clarified how do you deal with it. So all tiles share the same ETag for the same URL file?

Reply:
You should store the ETag returned on the initial header/root directory request, and pass that in the `If-Match` header on every subsequent request. If you get a `412 Precondition Failed` response that means the file has changed on storage, all byte ranges are invalid, and you can invalidate any entries in the cache for that URL/ETag. I suppose you can also wait for those to fall out of LRU cache, if the ETag is part of the key.

In the case where the storage does not return a proper `ETag` you can store empty string, that would mean the tileset breaks when changed on storage but there's nothing we can do client-side to detect that.